wire-ios-request-strategy/Sources/Request Strategies/User Clients/FetchingClientRequestStrategy.swift

```diff
@@ -110,13 +110,22 @@ public final class FetchingClientRequestStrategy: AbstractRequestStrategy {
     }
 
     public override func nextRequestIfAllowed(for apiVersion: APIVersion) -> ZMTransportRequest? {
+        // There may exist some clients that need an update, so try to sync any before asking
+        // for requests.
+        syncClientsNeedingUpdateIfNeeded()
+
         return
             userClientsByUserClientID.nextRequest(for: apiVersion) ??
             userClientsByUserID.nextRequest(for: apiVersion) ??
             userClientsByQualifiedUserID.nextRequest(for: apiVersion) ??
             entitySync.nextRequest(for: apiVersion)
     }
 
+    private func syncClientsNeedingUpdateIfNeeded() {
+        let clients = UserClient.fetchClientsNeedingUpdateFromBackend(in: managedObjectContext)
+        fetch(userClients: clients)
```

Comment:
I assume this does the syncing but this is async, right, we don't wait for it to be finished?
Is that an issue?

Reply:
It extracts the client ids and enqueues them the appropriate "sync" object, which is what generates a request. For instance, if I add client A to the `useClientsByClientID` sync object, then if we ask it for a next request it will produce one.

Reply:
The flow is something like:

- generate request for new message
- new message is dependent on new client
- generate request for new client
- process response for new client
- generate request for new message (succeeds this time because new message has no dependencies anymore)
- process response for new message

Reply:
Do we get a response or an error code from the first generated request in your flow?

Reply:
No, actually no request is generated. When we ask to generate a request (from the dependency sync) it will not return one because the dependency needs to be dealt with first.

Reply:
ok cool, so the order is respected, thks